tls/s2n_psk.c

```diff
@@ -24,49 +24,75 @@
 #include "utils/s2n_safety.h"
 
 #define S2N_HASH_ALG_COUNT S2N_HASH_SENTINEL
+#define S2N_MAX_OFFERED_PSK_LIST_SIZE (UINT16_MAX - sizeof(uint16_t) /* extension type */ - sizeof(uint16_t) /* extension size */)
 
-int s2n_psk_init(struct s2n_psk *psk, s2n_psk_type type)
+S2N_RESULT s2n_psk_init(struct s2n_psk *psk, s2n_psk_type type)
 {
-    notnull_check(psk);
+    ENSURE_MUT(psk);
 
-    memset_check(psk, 0, sizeof(struct s2n_psk));
+    CHECKED_MEMSET(psk, 0, sizeof(struct s2n_psk));
     psk->hmac_alg = S2N_HMAC_SHA256;
     psk->type = type;
 
-    return S2N_SUCCESS;
+    return S2N_RESULT_OK;
+}
+
+struct s2n_psk* s2n_external_psk_new()
+{
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    GUARD_PTR(s2n_alloc(&mem, sizeof(struct s2n_psk)));
+
+    struct s2n_psk *psk = (struct s2n_psk*)(void*) mem.data;
+    GUARD_RESULT_PTR(s2n_psk_init(psk, S2N_PSK_TYPE_EXTERNAL));
+
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return psk;
 }
 
-int s2n_psk_new_identity(struct s2n_psk *psk, const uint8_t *identity, size_t identity_size)
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size)
 {
     notnull_check(psk);
+    notnull_check(identity);
+    ENSURE_POSIX(identity_size != 0, S2N_ERR_INVALID_ARGUMENT);
 
     GUARD(s2n_realloc(&psk->identity, identity_size));
     memcpy_check(psk->identity.data, identity, identity_size);
 
     return S2N_SUCCESS;
 }
 
-int s2n_psk_new_secret(struct s2n_psk *psk, const uint8_t *secret, size_t secret_size)
+int s2n_psk_set_secret(struct s2n_psk *psk, const uint8_t *secret, uint16_t secret_size)
 {
     notnull_check(psk);
+    notnull_check(secret);
+    ENSURE_POSIX(secret_size != 0, S2N_ERR_INVALID_ARGUMENT);
 
     GUARD(s2n_realloc(&psk->secret, secret_size));
     memcpy_check(psk->secret.data, secret, secret_size);
 
     return S2N_SUCCESS;
 }
 
-int s2n_psk_free(struct s2n_psk *psk)
+S2N_CLEANUP_RESULT s2n_psk_wipe(struct s2n_psk *psk)
 {
     if (psk == NULL) {
-        return S2N_SUCCESS;
+        return S2N_RESULT_OK;
     }
 
-    GUARD(s2n_free(&psk->early_secret));
-    GUARD(s2n_free(&psk->identity));
-    GUARD(s2n_free(&psk->secret));
+    GUARD_AS_RESULT(s2n_free(&psk->early_secret));
+    GUARD_AS_RESULT(s2n_free(&psk->identity));
+    GUARD_AS_RESULT(s2n_free(&psk->secret));
```

Comment:
Doesn't wipe usually retain the allocated memory?

Reply:
The behavior of _wipe varies. Sometimes it clears memory, sometimes it reuses it.